File: items.cpp
/**
* @file     items.cpp
* @brief    功能项函数文件
* @details  主要包含功能项类的实现
*           items表示StimulationScene中的功能项形状。
*           它继承了QGraphicsPolygonItem，并为每个形状提供了一个多边形。枚举ItemType对每个流程图形状都有一个值。
* @author   LIN Guocheng
* @date     2021-8-9
* @version  0.0.1
**********************************************************************************
* @attention
* QT版本：5.12.11
* @par 修改日志:
* <table>
* <tr><th>Date        <th>Version  <th>Author    <th>Description
* <tr><td>2021/08/06  <td>0.0.1    <td>LIN Guocheng  <td>创建初始版本
* </table>
*
**********************************************************************************
*/

#include "items.h"
#include "arrow.h"

#include <QGraphicsScene>
#include <QGraphicsSceneContextMenuEvent>
#include <QMenu>
#include <QPainter>

Items::Items(ItemType itemType, QMenu *contextMenu, QGraphicsItem *parent)
    :QGraphicsPolygonItem(parent)
{
    myItemType = itemType;
    myContextMenu = contextMenu;

    switch (myItemType) {
        case test:
            myPolygon << QPointF(-100, -100) << QPointF(100, -100)
                      << QPointF(100, 100) << QPointF(-100, 100)
                      << QPointF(-100, -100);
            break;
        default:
            myPolygon << QPointF(-120, -80) << QPointF(-70, 80)
                      << QPointF(120, 80) << QPointF(70, -80)
                      << QPointF(-120, -80);
            break;
    }
    setPolygon(myPolygon);
    setFlag(QGraphicsItem::ItemIsMovable, true);    // 设置项可移动
    setFlag(QGraphicsItem::ItemIsSelectable, true); // 设置项可选中
    setFlag(QGraphicsItem::ItemSendsGeometryChanges, true); // 设置通知可发送
}

/**
 * @brief Items::removeArrow    当与连接线相关的项被删除时，删除该连接线
 * @param arrow Arrow
 */
void Items::removeArrow(Arrow *arrow)
{
    int index = arrows.indexOf(arrow);
    if(index != -1)
        arrows.removeAt(index);
}

/**
 * @brief Items::removeArrows   当一个项被删除时，删除与这个项有关的所有连接线
 */
void Items::removeArrows()
{
    foreach (Arrow *arrow, arrows)
    {
        arrow->startItem()->removeArrow(arrow); // 起始项删除箭头
        arrow->endItem()->removeArrow(arrow);   // 终止项删除箭头
        scene()->removeItem(arrow); // 删除箭头项
        delete arrow;
    }
}

/**
 * @brief Items::addArrow   将箭头添加到箭头列表
 * @param arrow 要添加的箭头Arrow
 */
void Items::addArrow(Arrow *arrow)
{
    arrows.append(arrow);
}

/**
 * @brief Items::image  为工具箱中的工具按钮创建图标
 * @return QPixmap
 */
QPixmap Items::image() const
{
    QPixmap pixmap(250, 250);
    pixmap.fill(Qt::transparent);
    QPainter painter(&pixmap);
    painter.setPen(QPen(Qt::black, 8));
    painter.translate(125, 125);
    painter.drawPolyline(myPolygon);

    return pixmap;
}

/**
 * @brief Items::contextMenuEvent   鼠标右键时显示菜单，默认情况下不选择项目
 * @param event QGraphicsSceneContextMenuEvent
 */
void Items::contextMenuEvent(QGraphicsSceneContextMenuEvent *event)
{
    scene()->clearSelection();  // 清除其他项的选中状态
    setSelected(true);  // 此项可选，则该项为选中状态
    myContextMenu->exec(event->screenPos());    // 在事件发生的地方显示菜单
}

/**
 * @brief Items::itemChange 如果项移动，更新箭头连接的位置
 * @param change    项移动事件GraphicsItemChange
 * @param value     QVariant
 * @return 箭头连接线更新结果
 */
QVariant Items::itemChange(GraphicsItemChange change, const QVariant &value)
{
    if(change == QGraphicsItem::ItemPositionChange)
    {
        foreach(Arrow *arrow, arrows)
        {
            arrow->updatePosition();
        }
    }
    return value;
}

File: items.py
from PySide6.QtCore import (QPointF, Qt, QRectF)
from PySide6.QtGui import (QPainter, QPainterPath, QPen, QPixmap, QPolygonF, QImage, QFont)
from PySide6.QtWidgets import (QGraphicsItem, QGraphicsPolygonItem)
from nrzook import nrz_ook
from rz33ook import rz33_ook
from rz50ook import rz50_ook
from rz66ook import rz66_ook

class Items(QGraphicsPolygonItem):
    NRZOOK, RZ33OOk, RZ50OOK, RZ66OOK, TestItem = range(5)

    def __init__(self, item_type, contextMenu, parent=None, scene=None):
        super().__init__(parent, scene)
        
        self.__inputNum = 0     # 输入值
        self.__outputNum = 0    #输出值
        self.__E0 = 3   # 调制器信号幅度
        self.__fc = 4   # 调制器信号频率
        
        self.arrows = []        
        self.item_type = item_type
        self._my_context_menu = contextMenu

        path = QPainterPath()
        if self.item_type == self.NRZOOK:
            
#            self._my_polygon = QPolygonF([
#                    QPointF(-100/2, -100/2), QPointF(100/2, -100/2),
#                    QPointF(100/2, 100/2), QPointF(-100/2, 100/2),
#                    QPointF(-100/2, -100/2)])
            path = QPainterPath()
            timesFont = QFont("Times", 30)
            timesFont.setStyleStrategy(QFont.ForceOutline)
            path.addText(-65, 25, timesFont, "NRZ-OOK")
#            path.moveTo(-100/2, -100.0/2)
#            path.lineTo(100.0/2, -100.0/2)
#            path.lineTo(100.0/2, 100.0/2)
#            path.lineTo(-100.0/2, 100.0/2)
            path.closeSubpath()
            path.setFillRule(Qt.WindingFill)
            self._my_polygon = path.toFillPolygon()
    
            #self.img = QImage(':/images/background1.png')
    
        elif self.item_type == self.TestItem:
            path = QPainterPath()
            timesFont = QFont("Times", 30)
            timesFont.setStyleStrategy(QFont.ForceOutline)
            path.addText(-65, 25, timesFont, "乘方器")
            path.closeSubpath()
            path.setFillRule(Qt.WindingFill)
            self._my_polygon = path.toFillPolygon()
            #self.img = QImage(':/images/power.png')
        
        elif self.item_type == self.RZ33OOk:
            path = QPainterPath()
            timesFont = QFont("Times", 30)
            timesFont.setStyleStrategy(QFont.ForceOutline)
            path.addText(-65, 25, timesFont, "RZ33-OOk")
            path.closeSubpath()
            path.setFillRule(Qt.WindingFill)
            self._my_polygon = path.toFillPolygon()
            
        elif self.item_type == self.RZ50OOK:
            path = QPainterPath()
            timesFont = QFont("Times", 30)
            timesFont.setStyleStrategy(QFont.ForceOutline)
            path.addText(-65, 25, timesFont, "RZ50-OOK")
            path.closeSubpath()
            path.setFillRule(Qt.WindingFill)
            self._my_polygon = path.toFillPolygon()
            
        elif self.item_type == self.RZ66OOK:
            path = QPainterPath()
            timesFont = QFont("Times", 30)
            timesFont.setStyleStrategy(QFont.ForceOutline)
            path.addText(-65, 25, timesFont, "RZ66-OOK")
            path.closeSubpath()
            path.setFillRule(Qt.WindingFill)
            self._my_polygon = path.toFillPolygon()
            #self.img = QImage(':/images/power.png')
        else:
            
            self._my_polygon = QPolygonF([
                    QPointF(-120/2, -80/2), QPointF(-70/2, 80/2),
                    QPointF(120/2, 80/2), QPointF(70/2, -80/2),
                    QPointF(-120/2, -80/2)])
            
            #self.img = QImage(':/images/background1.png')
        #self.setPolygon(self.img)
        #self.setImg(self.img)
        self.setPolygon(self._my_polygon)
        self.setFlag(QGraphicsItem.ItemIsMovable, True)
        self.setFlag(QGraphicsItem.ItemIsSelectable, True)
    
    # 删除单个连接线
    def remove_arrow(self, arrow):
        try:
            self.arrows.remove(arrow)
        except ValueError:
            pass

    # 删除多个连接线
    def remove_arrows(self):
        for arrow in self.arrows[:]:
            arrow.start_item().remove_arrow(arrow)
            arrow.end_item().remove_arrow(arrow)
            self.scene().removeItem(arrow)

    # 添加连接线
    def add_arrow(self, arrow):
        self.arrows.append(arrow)

    # 图标
    
    def image(self):
        
        pixmap = QPixmap(250, 250)
        pixmap.fill(Qt.transparent)
        painter = QPainter(pixmap)
        painter.setPen(QPen(Qt.black, 8))
        painter.translate(125, 125)
        painter.drawPolyline(self._my_polygon)
        #rect = QRectF(200,100,self.img.width()/2,self.img.height()/2)
        #painter.drawImage(rect, self.img)
        #pixmap = QPixmap(self.img)
        return pixmap
        
        #img = self.getImgUrl()
        #return img
    
    # 菜单栏
    def contextMenuEvent(self, event):
        self.scene().clearSelection()
        self.setSelected(True)
        self._my_context_menu.exec(event.screenPos())
        
    # 当项目被选中时发出信号
    def itemChange(self, change, value):
        if change == QGraphicsItem.ItemPositionChange:
            for arrow in self.arrows:
                arrow.updatePosition()

        return value
    

    
    # 计算结果
    def calculateResult(self, itemType, inputNum):
        if(itemType == self.TestItem):
            outputNum = inputNum * inputNum
            self.__inputNum = inputNum
            self.__outputNum = outputNum
            print("TestItem:")
            print("inputNum = ", self.__inputNum)
            print("outputNum = ", self.__outputNum)            
        elif(itemType == self.NRZOOK):
            nrz = nrz_ook(self.getE0(), self.getfc())
            outputNum = nrz.E_out(inputNum)
            self.__inputNum = inputNum
            self.__outputNum =  outputNum
            print("NRZOOK:")
            print("inputNum = ", self.__inputNum)
            print("outputNum = ", outputNum.real, "+j", outputNum.imag)
        elif(itemType == self.RZ33OOk):
            rz33 = rz33_ook(self.getE0(), self.getfc())
            outputNum = rz33.E_out(inputNum)
            self.__inputNum = inputNum
            self.__outputNum =  outputNum
            print("RZ33OOk:")
            print("inputNum = ", self.__inputNum)
            print("outputNum = ", outputNum.real, "+j", outputNum.imag)
        elif(itemType == self.RZ50OOK):
            rz50 = rz50_ook(self.getE0(), self.getfc())
            outputNum = rz50.E_out(inputNum)
            self.__inputNum = inputNum
            self.__outputNum =  outputNum
            print("RZ50OOK:")
            print("inputNum = ", self.__inputNum)
            print("outputNum = ", outputNum.real, "+j", outputNum.imag)
        elif(itemType == self.RZ66OOK):
            rz66 = rz66_ook(self.getE0(), self.getfc())
            outputNum = rz66.E_out(inputNum)
            self.__inputNum = inputNum
            self.__outputNum =  outputNum
            print("RZ66OOK:")
            print("inputNum = ", self.__inputNum)
            print("outputNum = ", outputNum.real, "+j", outputNum.imag)
        return outputNum            
        
    
##################SET-GET METHOD##############################
    # 设置输入值
    def setInputNum(self, inputNum):
        self.__inputNum = inputNum
    
    # 得到输入值
    def getInputNum(self):
        return self.__inputNum
    
    # 设置输出值
    def setOutputNum(self, outputNum):
        self.__outputNum = outputNum
        
    # 得到输出值
    def getOutputNum(self):
        return self.__outputNum
    
    # 设置调制器幅度
    def setE0(self, E0):
        self.__E0 = E0
    
    # 得到调制器幅度
    def getE0(self):
        return self.__E0
    
    # 设置调制器频率
    def setfc(self, fc):
        self.__fc = fc
        
    # 得到调制器频率
    def getfc(self):
        return self.__fc
    
    # 设置输入值
    def setImg(self, imgUrl):
        self.imgUrl = imgUrl
    
    # 得到输入值
    def getImg(self):
        return self.imgUrl
##################SET-GET METHOD##############################